src/Streamly/Internal/Data/Stream/IsStream/Top.hs

```diff
@@ -0,0 +1,571 @@
+{-# INLINE leftJoin #-}
+leftJoin :: Monad m =>
+    (a -> b -> Bool) -> SerialT m a -> SerialT m b -> SerialT m (a, Maybe b)
+leftJoin eq s1 s2 = Stream.evalStateT (return False) $ do
+    a <- Stream.liftInner s1
+    -- XXX should we use StreamD monad here?
+    -- XXX Is there a better way to perform some action at the end of a loop
+    -- iteration?
+    lift $ put False
+    let final = do
+            r <- lift get
+            if r
+            then StreamK.nil
+            else StreamK.yield Nothing
+    b <- fmap Just (Stream.liftInner s2) <> final
+    case b of
+        Just b1 ->
+            if a `eq` b1
+            then do
+                lift $ put True
+                return (a, Just b1)
+            else StreamK.nil
+        Nothing -> return (a, Nothing)
+
+-- | Like 'outerJoin' but uses a hashmap for efficiency.
+--
+-- Space: O(n)
+--
+-- Time: O(m + n)
+--
+-- /Unimplemented/
+{-# INLINE hashLeftJoin #-}
+hashLeftJoin :: -- Hashable b =>
+    (a -> b -> Bool) -> t m a -> t m b -> t m (a, Maybe b)
+hashLeftJoin = undefined
+
+-- | Like 'leftJoin' but works on sorted streams.
+--
+-- Space: O(1)
+--
+-- Time: O(m + n)
+--
+-- /Unimplemented/
+{-# INLINE mergeLeftJoin #-}
+mergeLeftJoin :: -- Monad m =>
+    (a -> b -> Ordering) -> t m a -> t m b -> t m (a, Maybe b)
+mergeLeftJoin _eq _s1 _s2 = undefined
+
+-- XXX We can do this concurrently.
+--
+-- | For all elements in @t m a@, for all elements in @t m b@ if @a@ and @b@
+-- are equal by the given equality pedicate then return the tuple (Just a, Just
+-- b).  If @a@ is not found in @t m b@ then return (a, Nothing), return
+-- (Nothing, b) for vice-versa.
+--
+-- For space efficiency use the smaller stream as the second stream.
+--
+-- Space: O(n)
+--
+-- Time: O(m x n)
+--
+-- /Unimplemented/
+{-# INLINE outerJoin #-}
+outerJoin :: MonadIO m =>
+       (a -> b -> Bool)
+    -> SerialT m a
+    -> SerialT m b
+    -> SerialT m (Maybe a, Maybe b)
+outerJoin eq s1 s =
+    Stream.concatM $ do
+        arr <- Array.fromStream $ fmap (,False) s
+        return $ go arr <> leftOver arr
+
+    where
+
+    leftOver =
+            fmap (\(x, _) -> (Nothing, Just x))
+                . Stream.filter (not . snd)
+                . Array.toStream
+
+    go arr = Stream.evalStateT (return False) $ do
+        a <- Stream.liftInner s1
+        -- XXX should we use StreamD monad here?
+        -- XXX Is there a better way to perform some action at the end of a loop
+        -- iteration?
+        lift $ put False
+        let final = do
+                r <- lift get
+                if r
+                then StreamK.nil
+                else StreamK.yield Nothing
+        (_i, b) <-
+            Stream.indexed
+                $ fmap Just (Stream.liftInner (Array.toStream arr)) <> final
+        case b of
+            Just (b1, _used) ->
+                if a `eq` b1
+                then do
+                    lift $ put True
+                    -- XXX Need to use a mutable array
```

Comment:
This function is currently incorrectly implemented?

Reply:
Yes, this todo needs to be done. The doc for the function still says "unimplemented".